daemon/logger/jsonfilelog/jsonfilelog.go

```diff
@@ -59,15 +86,98 @@ func (l *JSONFileLogger) Log(msg *logger.Message) error {
 		return err
 	}
 	l.buf.WriteByte('\n')
-	_, err = l.buf.WriteTo(l.f)
+	_, err = writeLog(l)
+	return err
+}
+
+func writeLog(l *JSONFileLogger) (int64, error) {
+	if l.capacity == -1 {
+		return writeToBuf(l)
+	}
+	meta, err := l.f.Stat()
+	if err != nil {
+		return -1, err
+	}
+	if meta.Size() >= l.capacity {
+		name := l.f.Name()
+		if err := l.f.Close(); err != nil {
+			return -1, err
+		}
+		if err := rotate(name, l.n); err != nil {
+			return -1, err
+		}
+		file, err := os.OpenFile(name, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, 0666)
+		if err != nil {
+			return -1, err
+		}
+		l.f = file
+	}
+	return writeToBuf(l)
+}
+
+func writeToBuf(l *JSONFileLogger) (int64, error) {
+	i, err := l.buf.WriteTo(l.f)
 	if err != nil {
-		// this buffer is screwed, replace it with another to avoid races
 		l.buf = bytes.NewBuffer(nil)
+	}
+	return i, err
+}
+
+func rotate(name string, n int) error {
+	if n < 2 {
+		return nil
+	}
+	for i := n - 1; i > 1; i-- {
+		oldFile := name + "." + strconv.Itoa(i)
+		replacingFile := name + "." + strconv.Itoa(i-1)
+		if err := backup(oldFile, replacingFile); err != nil {
+			return err
+		}
+	}
+	if err := backup(name+".1", name); err != nil {
 		return err
 	}
 	return nil
 }
 
+func backup(old, curr string) error {
+	if _, err := os.Stat(old); !os.IsNotExist(err) {
+		err := os.Remove(old)
+		if err != nil {
+			return err
+		}
+	}
+	if _, err := os.Stat(curr); os.IsNotExist(err) {
+		if f, err := os.Create(curr); err != nil {
+			return err
+		} else {
+			f.Close()
+		}
+	}
+	return os.Rename(curr, old)
+}
+
+func ValidateLogOpt(cfg map[string]string) error {
+	for key := range cfg {
+		switch key {
+		case "max-file":
+		case "max-size":
+		default:
+			return fmt.Errorf("unknown log opt '%s' for json-file log driver", key)
+		}
+	}
+	return nil
+}
+
+func (l *JSONFileLogger) GetReaderByIndex(index int) (io.Reader, error) {
```

Comment:
Omg, I don't like this so much. But I think we have no choice now. `GetReader` was superbad idea in the first place.
I don't know, maybe use type assertion instead of bringing this to interface?